carbon-design-system/carbon-for-ibm-dotcom
# fix(search-with-typeahead): make eventBeforeRedirect return consistent & useful data

### Related Ticket(s)

Closes https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/11256

### Description

The reporter of the issue was using a demo I'd set up showing how to customize the masthead's search-with-typeahead. While the `null` error is a bug in my demo, it reveals an issue: the `c4d-search-with-typeahead-beingredirected` event can be dispatched from two places, and the `details` returned are not consistent in these two places. Additionally, the same event can be dispatched twice in rapid succession if a listener calls `event.preventDefault()`.

This PR makes the returned details consistent, and it improves them by giving listeners easy access to what they're probably looking for anyway: the search query.

### Changelog

**New**

- Passes the current search query, be that the text input value or a typeahead item selected, to `c4d-search-with-typeahead-beingredirected` event listeners.

**Changed**

- Improves readability of `search-with-typeahead` component by extracting some logic into a standalone method and making consistent use of getters.

### Testing

Masthead deploy preview Story: https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/11376/index.html?path=/story/components-masthead--default

You'll need to verify that both searching by a query string and selecting typeahead result items behave consistently with regards to the `c4d-search-with-typeahead-beingredirected` event.

1. Add an event listener for the event using the browser console, e.g.:
   ```
   document.documentElement.addEventListener(
     'c4d-search-with-typeahead-beingredirected',
     (e) => {
       e.preventDefault();
       const { redirectUrl, searchQueryString } = e.detail;
       const newUrl = new URL(`https://duckduckgo.com/?q=${searchQueryString}`).toString();
       debugger;
       alert(`At this point, you could navigate to ${newUrl}`);
     }
   );
   ```
2. In the debugger, verify the `redirectUrl` and `searchQueryString` event details return the correct information for each of the user inputs.
3. Verify the event is not dispatched twice by confirming only one alert appears per search action.
